pkg/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go

```diff
@@ -264,17 +270,51 @@ func (n *klusterletController) sync(ctx context.Context, controllerContext facto
 		}))
 		return err
 	}
-	// TODO store this in the status of the klusterlet itself
-	n.workGeneration = generation
+	observedKlusterletGeneration := klusterlet.Generation
 
 	// if we get here, we have successfully applied everything and should indicate that
-	helpers.UpdateKlusterletStatus(ctx, n.klusterletClient, klusterletName, helpers.UpdateKlusterletConditionFn(operatorapiv1.StatusCondition{
-		Type: klusterletApplied, Status: metav1.ConditionTrue, Reason: "KlusterletApplied",
-		Message: "Klusterlet Component Applied",
-	}))
+	helpers.UpdateKlusterletStatus(ctx, n.klusterletClient, klusterletName,
+		helpers.UpdateKlusterletConditionFn(operatorapiv1.StatusCondition{
+			Type: klusterletApplied, Status: metav1.ConditionTrue, Reason: "KlusterletApplied",
+			Message: "Klusterlet Component Applied"}),
+		helpers.UpdateKlusterletGenerationsFn(registrationGeneration, workGeneration),
+		func(oldStatus *operatorapiv1.KlusterletStatus) error {
+			oldStatus.ObservedGeneration = observedKlusterletGeneration
+			return nil
+		},
+	)
 	return nil
 }
 
+// Check if cluster name is changed in hub kubeconfig secret by comparing with deployment args. Force update the deployment
+// if the secret is changed.
+func (n *klusterletController) forceRollOutWorkAgent(ctx context.Context, config klusterletConfig, forceRollOut *bool) {
```

Comment:
My own personal preference here is to avoid mutating inputs and instead take an input parameter and output the result. Think it over, it can be done in a follow-up.